internal/monitors/statsd/statsd.go

```diff
@@ -0,0 +1,148 @@
+package statsd
+
+import (
+	"errors"
+	"net"
+	"strconv"
+	"strings"
+)
+
+type statsDListener struct {
+	ipAddr     string
+	port       uint16
+	protocol   string
+	udpConn    *net.UDPConn
+	tcpConn    *net.TCPConn
+	prefix     string
+	converters []*converter
+}
+
+type statsDMetric struct {
+	metricName string
+	metricType string
+	value      int64
+	dimensions map[string]string
+}
+
+func (sl *statsDListener) Listen() error {
+	switch sl.protocol {
+	case "tcp":
+		return sl.listenTCP()
+	case "udp":
+		return sl.listenUDP()
+	}
+
+	return errors.New("Invalid protocol")
+}
+
+func (sl *statsDListener) listenUDP() error {
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{
+		IP:   net.ParseIP(sl.ipAddr),
+		Port: int(sl.port),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	sl.udpConn = conn
+	return nil
+}
+
+func (sl *statsDListener) listenTCP() error {
+	return nil
+}
+
+func (sl *statsDListener) FetchMetrics() ([]*statsDMetric, error) {
+	var rawMetrics []string
+	var err error
+
+	switch sl.protocol {
+	case "tcp":
+		rawMetrics, err = sl.readMetricsTCP()
+	case "udp":
+		rawMetrics, err = sl.readMetricsUDP()
+	default:
+		return nil, errors.New("Connection has not established")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	parsed := parseMetrics(rawMetrics, sl.converters, sl.prefix)
+
+	return parsed, nil
+}
+
+func (sl *statsDListener) readMetricsTCP() ([]string, error) {
+	return nil, nil
+}
+
+func (sl *statsDListener) readMetricsUDP() ([]string, error) {
+	// UDP needs to receive data packet by packet. Max packet size is 65535 for now.
+	buf := make([]byte, 65536)
+	n, _, err := sl.udpConn.ReadFromUDP(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(string(buf[0:n]), "\n"), nil
+}
+
+func parseMetrics(raw []string, converters []*converter, prefix string) []*statsDMetric {
+	var metrics []*statsDMetric
+
+	for _, m := range raw {
+		colonIdx := strings.Index(m, ":")
+		pipeIdx := strings.Index(m, "|")
+
+		rawMetricName := m[0:colonIdx]
+		metricName := ""
+
+		var dims map[string]string
+
+		if converters == nil {
+			metricName = rawMetricName
+		} else {
+			var dimSrc string
+
+			if prefix != "" {
+				dimSrc = removePrefix(rawMetricName, prefix)
+			} else {
+				dimSrc = rawMetricName
+			}
+
+			metricName, dims = convertMetric(dimSrc, converters)
+		}
+
+		strValue := m[colonIdx+1 : pipeIdx]
+		value, err := strconv.ParseInt(strValue, 10, 64)
+
+		if err == nil {
+			metrics = append(metrics, &statsDMetric{
+				metricName: metricName,
+				metricType: m[pipeIdx+1:],
```

Comment:
There is an optional component after the metric type that specified the rate.  See https://github.com/statsd/statsd/blob/master/docs/metric_types.md#sampling.  We don't have to support this but it should at least parse it right.